pkg/providers/az-client.go

```diff
@@ -1,32 +1,138 @@
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
 	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/subscription/armsubscription"
-	"github.com/Azure/draft/pkg/cred"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
 )
 
+//go:generate mockgen -source=./az-client.go -destination=./mock/az-client.go .
+type AzClientInterface interface {
+	ListResourceGroups(ctx context.Context, subscriptionID string) ([]armresources.ResourceGroup, error)
+	ListTenants(ctx context.Context) ([]armsubscription.TenantIDDescription, error)
+	assignSpRole(ctx context.Context, subscriptionId, resourceGroup, servicePrincipalObjectID, roleId string) error
+}
+
+// assert AzClient implements AzClientInterface
+var _ AzClientInterface = &AzClient{}
+
+// AzClient is a struct that contains the Azure client and its dependencies
+// It is used to interact with Azure resources
+// Create a new AzClient with NewAzClient
 type AzClient struct {
-	AzTenantClient   azTenantClient
-	RoleAssignClient *armauthorization.RoleAssignmentsClient
+	Credential          *azidentity.DefaultAzureCredential
+	TenantClient        *armsubscription.TenantsClient
+	RoleAssignClient    *armauthorization.RoleAssignmentsClient
+	ResourceGroupClient *armresources.ResourceGroupsClient
 }
 
-//go:generate mockgen -source=./az-client.go -destination=./mock/az-client.go .
-type azTenantClient interface {
-	NewListPager(options *armsubscription.TenantsClientListOptions) *runtime.Pager[armsubscription.TenantsClientListResponse]
+func NewAzClient(cred *azidentity.DefaultAzureCredential) (*AzClient, error) {
+	azClient := &AzClient{
+		Credential: cred,
+	}
+	return azClient, nil
 }
 
-func createRoleAssignmentClient(subscriptionId string) (*armauthorization.RoleAssignmentsClient, error) {
-	credential, err := cred.GetCred()
-	if err != nil {
-		return nil, fmt.Errorf("failed to get credentials: %w", err)
+func (az *AzClient) ListResourceGroups(ctx context.Context, subscriptionID string) ([]armresources.ResourceGroup, error) {
+	log.Debug("listing Azure resource groups for subscription ", subscriptionID)
+	if az.ResourceGroupClient == nil {
+		c, err := armresources.NewResourceGroupsClient(subscriptionID, az.Credential, nil)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create resource group client: %w", err)
+		}
+		az.ResourceGroupClient = c
+	}
+
+	var rgs []armresources.ResourceGroup
+	pager := az.ResourceGroupClient.NewListPager(nil)
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("listing resource groups page: %w", err)
+		}
+
+		for _, rg := range page.Value {
+			if rg == nil {
+				return nil, errors.New("nil rg")
+			}
+
+			rgs = append(rgs, *rg)
+		}
+	}
+
+	return rgs, nil
+}
+
+func (az *AzClient) ListTenants(ctx context.Context) ([]armsubscription.TenantIDDescription, error) {
+	log.Debug("listing Azure subscriptions")
+
+	var tenants []armsubscription.TenantIDDescription
```

Comment:
💡 minor: How about use `make([]armsubscription.TenantIDDescription, 0)` to initialize the tenants slice.

```*.go
       // Initialize the tenant slice to store the results.
	tenants := make([]armsubscription.TenantIDDescription, 0)
```